components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/filter/FilterTreeBuilder.java

```diff
@@ -0,0 +1,285 @@
+    }
+
+    /**
+     * Builds the binary tree from the filterString.
+     *
+     * @return the filter tree as root.
+     */
+    public Node buildTree() throws OrganizationManagementClientException {
+
+        expression();
+        return root;
+    }
+
+    /**
+     * We build the parser using the recursive descent parser technique.
+     */
+    private void expression() throws OrganizationManagementClientException {
+
+        term();
+        while (symbol.equals(OrganizationManagementConstants.Filter.OR)) {
+            OperationNode or = new OperationNode(OrganizationManagementConstants.Filter.OR);
+            or.setLeftNode(root);
+            term();
+            or.setRightNode(root);
+            root = or;
+        }
+    }
+
+    /**
+     * We build the parser using the recursive descent parser technique.
+     */
+    private void term() throws OrganizationManagementClientException {
+
+        factor();
+        while (symbol.equals(OrganizationManagementConstants.Filter.AND)) {
+            OperationNode and = new OperationNode(OrganizationManagementConstants.Filter.AND);
+            and.setLeftNode(root);
+            factor();
+            and.setRightNode(root);
+            root = and;
+        }
+    }
+
+    /**
+     * We build the parser using the recursive descent parser technique.
+     */
+    private void factor() throws OrganizationManagementClientException {
+
+        symbol = nextSymbol();
+        if (symbol.equals(OrganizationManagementConstants.Filter.NOT)) {
+            OperationNode not = new OperationNode(OrganizationManagementConstants.Filter.NOT);
+            factor();
+            not.setRightNode(root);
+            root = not;
+        } else if (symbol.equals("(")) {
+            expression();
+            symbol = nextSymbol(); // We don't care about ')'.
+        } else {
+            if (!(symbol.equals(")"))) {
+                ExpressionNode expressionNode = new ExpressionNode();
+                validateAndBuildFilterExpression(symbol, expressionNode);
+                root = expressionNode;
+                symbol = nextSymbol();
+            } else {
+                throw handleClientException(ERROR_CODE_INVALID_FILTER_ARGUMENT, symbol);
+            }
+        }
+    }
+
+    /**
+     * Validate the simple filter and build a ExpressionNode
+     *
+     * @param filterString   the filter string.
+     * @param expressionNode the expression node.
+     */
+    private void validateAndBuildFilterExpression(String filterString, ExpressionNode expressionNode)
+            throws OrganizationManagementClientException {
+
+        if (StringUtils.isNotBlank(filterString) && !filterString.equals("-1")) {
+            String trimmedFilter = filterString.trim();
+            String[] filterParts;
+
+            if (Pattern.compile(Pattern.quote(" eq "), Pattern.CASE_INSENSITIVE).matcher(filterString).find()) {
+                filterParts = trimmedFilter.split(" eq | EQ | eQ | Eq ");
```

Comment:
do we need to support case insensitivity?

Reply:
Case insensitivity supported considering implementation in framework.